Source/WebCore/page/WebKitStringMatcher.cpp

```diff
@@ -0,0 +1,277 @@
+}
+
+WebKitStringMatcher::WebKitStringMatcher(Ref<SharedMemory>&& data, std::span<const State> states, std::span<const Transition> transitions)
+    : m_data(WTFMove(data))
+    , m_states(states)
+    , m_transitions(transitions)
+{
+}
+
+WebKitStringMatcher::~WebKitStringMatcher() = default;
+
+struct WebKitStringMatcher::IdentifierAndIndex {
+    uint16_t identifier { 0 };
+    size_t stringIndex { 0 };
+};
+
+auto WebKitStringMatcher::runDFAOnEachCharacter(const String& input, const WebKitStringMatcherInit& options) -> Vector<IdentifierAndIndex>
+{
+    Vector<IdentifierAndIndex> result;
+
+    StringView stringView(input);
+    size_t start = options.searchReverse ? input.length() - 1 : 0;
+    size_t end = options.searchReverse ? -1 : input.length();
+    size_t increment = options.searchReverse ? -1 : 1;
+
+    for (size_t i = start; i < end; i += increment) {
+        size_t stateIndex { 0 };
+        for (size_t stringIndex = i; stringIndex < end; stringIndex += increment) {
+            auto& state = m_states[stateIndex];
+            auto codeUnit = stringView.characterAt(stringIndex);
+            bool foundMatch { false };
+            for (uint16_t transitionIndex = state.transitionsBeginIndexOrIdentifier; transitionIndex < state.transitionsEndIndexOrSentinel; transitionIndex++) {
+                auto& transition = m_transitions[transitionIndex];
+                if (codeUnit == transition.codeUnitToCheck) {
+                    stateIndex = transition.stateIndexToTransitionToIfMatched;
+                    foundMatch = true;
+                    break;
+                }
+            }
+            if (!foundMatch)
+                break;
+            while (m_states[stateIndex].transitionsEndIndexOrSentinel == State::identifierSentinel) {
+                result.append({
+                    m_states[stateIndex].transitionsBeginIndexOrIdentifier,
+                    stringIndex
+                });
+                if (!options.matchAll)
+                    return result;
+                stateIndex++;
+            }
+        }
+    }
+
+    return result;
+}
+
+JSC::JSValue WebKitStringMatcher::match(JSC::JSGlobalObject& globalObject, const String& input, const WebKitStringMatcherInit& options)
+{
+    auto result = runDFAOnEachCharacter(input, options);
+    JSContextRef context = ::toRef(&globalObject);
+    Vector<JSValueRef> values;
+    for (auto& identifierAndIndex : result) {
+        std::array<JSValueRef, 2> array {
+            JSValueMakeNumber(context, identifierAndIndex.identifier),
+            JSValueMakeNumber(context, identifierAndIndex.stringIndex)
+        };
+        JSValueRef match = JSObjectMakeArray(context, array.size(), array.data(), nullptr);
+        if (!options.matchAll)
+            return ::toJS(&globalObject, match);
+        values.append(match);
+    }
+    return ::toJS(&globalObject, JSObjectMakeArray(context, values.size(), values.span().data(), nullptr));
+}
+
+struct Trie {
```

Comment:
I'd like to see all the guts of the matcher pulled out into its own file (WebKitStringMatcherTrie.h if you will) so that WebKitStringMatcher.h/cpp can just be the binding to it.